PROTOCOL.md

```diff
@@ -1908,14 +1910,21 @@ The response contains multiple lines:
 
 #### Request Body
 
-The request body should be a JSON string containing the following two optional fields:
+The request body should be a JSON string containing the following optional fields:
  
 - **predicateHints** (type: String, optional): a list of SQL boolean expressions using [a restricted subset of SQL](#sql-expressions-for-filtering), in a JSON array.
   - When it’s present, the server will use the provided predicates as a hint to apply the SQL predicates on the returned files.
     - Filtering files based on the SQL predicates is **BEST EFFORT**. The server may return files that don’t satisfy the predicates.
     - If the server fails to parse one of the SQL predicates, or fails to evaluate it, the server may skip it.
     - Predicate expressions are conjunctive (AND-ed together).
   - When it’s absent, the server will return all of files in the table.
+  - **These will be deprecated once all the client and server implementation move to using jsonPredicateHints below**.
+
+- **jsonPredicateHints** (type: String, optional): query predicates on partition columns specified using [a structured JSON format](#json-predicates-for-filtering), which is encoded using Base64 encoding.
```

Comment:
Can we avoid using Base64? We can make `jsonPredicateHints` be a string, and just convert the content to a json string, then set it. See the `stats` field in `file`. In addition, can we make it be a json array similar to `predicateHints`?

Another way to embed the json content is:
```
{
  "jsonPredicateHints": [
    {"op":"isNull","children":[{"op":"column","name":"hireDate","valueType":"date"}]},
    {"op":"isNull","children":[{"op":"column","name":"state","valueType":"string"}]}
  ],
  "limitHint": 1000,
  "version": 123
}
```

Reply:
Base64 encoding is a common practice for not trivial data types.
We url-encode a lot of our params as well.

I had trouble packing/unpacking json predicate tree through various hoops during my testing (for example, http query body did not like it).
Once I switch to base64 encoding, things worked like charm.

We will need to change our implementation in spark connector and UC's delta sharing server.
Let me know if you all feel strongly against base64 encoding.

Reply:
> (for example, http query body did not like it).

This is likely you are not using the right API? Which approach were you using? `stats` approach or embedded json?

Base64 encoding is not readable. Would be great to avoid that.

Reply:
ok, removed the base64 encoding as per offline discussion.